otherlibs/runtime_events/runtime_events_consumer.c

```diff
@@ -493,8 +493,9 @@ caml_runtime_events_read_poll(struct caml_runtime_events_cursor *cursor,
           break;
         case EV_LIFECYCLE:
           if (cursor->lifecycle) {
+            int64_t data = msg_length > 2 ? buf[2] : 0;
             if( !cursor->lifecycle(domain_num, callback_data, buf[1],
-                                    RUNTIME_EVENTS_ITEM_ID(header), buf[2]) ) {
+                                    RUNTIME_EVENTS_ITEM_ID(header), data) ) {
```

Comment:
There are other code paths in this routine which access `buf[2]` as well: `EV_COUNTER`, `EV_ALLOC`, `EV_USER_MSG_TYPE_INT` and the `default` case for user events.

Shouldn't all of them be similarly guarded with length checks?

Also, earlier, the `msg_length > RUNTIME_EVENTS_MAX_MSG_LENGTH` check to reject corrupted or ill-formed data should probably grow a `|| msg_length < 2` condition as well, to reject truncated data, and make sure `buf[1]` is always safe to access.

Reply:
Good points, I'll try to add a testcase that triggers the failure for those and use a similar fix.

Reply:
`buf[2]` is uninitialized only when `write_to_ring` is called with `event_length == 0`
`event_length == 0` for: `EV_BEGIN`, `EV_EXIT`, `EV_USER_MSG_TYPE_UNIT`, `EV_USER_MSG_TYPE_SPAN_BEGIN`, `EV_USER_MSG_TYPE_SPAN_END`; but `buf[2]` is not accessed in these cases. `EV_USER` with `EV_USER_MSG_TYPE_CUSTOM` has a length `>=1`.

So well behaved events don't need additional bounds checks. However corrupted events need them. I've written a test to reproduce that, and will update the PR (as discussed elsewhere, I'll open a separate PR for CI changes and the runtime events bugfix).

Reply:
We cannot reject all messages < 2, otherwise we also reject padding events. I've implemented a slightly extended version of this in (which works for padding events too):
> commit a58463df31f63eb8cef2d3933aa22f90a1d6bb7f
>Author: Edwin Török <edwin.torok@cloud.com>
>Date:   Sun Jul 14 16:50:12 2024 +0100
>     runtime_events_consumer: guard agains short msg_length